samples/matter/common/src/bt_nus_service.cpp

```diff
@@ -0,0 +1,224 @@
+	VerifyOrReturnError(bt_passkey_set(kDefaultPasskey) != 0, CHIP_ERROR_INTERNAL);
+#endif
+	VerifyOrReturnError(bt_conn_auth_cb_register(&sConnAuthCallbacks) == 0, CHIP_ERROR_INTERNAL);
+	VerifyOrReturnError(bt_conn_auth_info_cb_register(&sConnAuthInfoCallbacks) == 0, CHIP_ERROR_INTERNAL);
+	VerifyOrReturnError(bt_nus_init(&sNusCallbacks) == 0, CHIP_ERROR_INTERNAL);
+
+	return CHIP_NO_ERROR;
+}
+
+void NUSService::StartServer()
+{
+	VerifyOrReturn(!mIsStarted, ChipLogError(DeviceLayer, "NUS service was already started"));
+
+	PlatformMgr().LockChipStack();
+	CHIP_ERROR ret = BLEAdvertisingArbiter::InsertRequest(mAdvertisingRequest);
+	PlatformMgr().UnlockChipStack();
+
+	if (CHIP_NO_ERROR != ret) {
+		ChipLogError(DeviceLayer, "Could not start NUS service");
+		mIsStarted = false;
+		return;
+	}
+	mIsStarted = true;
+	ChipLogProgress(DeviceLayer, "NUS service started");
+}
+
+void NUSService::StopServer()
+{
+	VerifyOrReturn(mIsStarted);
+
+	PlatformMgr().LockChipStack();
+	BLEAdvertisingArbiter::CancelRequest(mAdvertisingRequest);
+	PlatformMgr().UnlockChipStack();
+
+	mIsStarted = false;
+	ChipLogProgress(DeviceLayer, "NUS service stopped");
+}
+
+void NUSService::RxCallback(struct bt_conn *conn, const uint8_t *const data, uint16_t len)
+{
+	VerifyOrReturn(bt_conn_get_security(GetNUSService().mBTConnection) >= BT_SECURITY_L2);
+	ChipLogDetail(DeviceLayer, "NUS received: %d bytes", len);
+	GetNUSService().DispatchCommand(reinterpret_cast<const char *>(data), len);
+}
+
+CHIP_ERROR NUSService::RegisterCommand(const char *const name, size_t length, CommandCallback callback, void *context)
+{
+	VerifyOrReturnError(name, CHIP_ERROR_INVALID_ARGUMENT);
+
+	struct Command newCommand {};
+	memcpy(newCommand.command, name, length);
+	newCommand.commandLen = length;
+	newCommand.callback = callback;
+	newCommand.context = context;
+	mCommandsList.push_back(newCommand);
+
+	return CHIP_NO_ERROR;
+}
+
+void NUSService::DispatchCommand(const char *const data, uint16_t len)
+{
+	for (auto c : mCommandsList) {
+		if (strncmp(data, c.command, len) == 0) {
+			if (c.callback) {
+				PlatformMgr().LockChipStack();
+				c.callback(c.context);
+				PlatformMgr().UnlockChipStack();
+			}
+			return;
+		}
+	}
+	ChipLogError(DeviceLayer, "NUS command unkown!");
+}
+
+CHIP_ERROR NUSService::SendData(const char *const data, size_t length)
+{
+	VerifyOrReturnError(mBTConnection, CHIP_ERROR_SENDING_BLOCKED);
+	VerifyOrReturnError(bt_conn_get_security(GetNUSService().mBTConnection) >= BT_SECURITY_L2,
+			    CHIP_ERROR_SENDING_BLOCKED);
+	VerifyOrReturnError(bt_nus_send(mBTConnection, reinterpret_cast<const uint8_t *const>(data), length) == 0,
+			    CHIP_ERROR_SENDING_BLOCKED);
+	return CHIP_NO_ERROR;
+}
+
+void NUSService::Connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		ChipLogError(DeviceLayer, "NUS Connection failed (err %u)", err);
+		return;
+	}
+
+	bt_conn_set_security(conn, BT_SECURITY_L3);
+
+	char addr[BT_ADDR_LE_STR_LEN];
```

Comment:
Everywhere where you generate the address string just for the logging purpose you can guard the code with `#if CHIP_<level>_LOGGING` to reduce the release builds. Level can be `PROGRESS` etc.